src/transformers/generation/utils.py

```diff
@@ -2315,6 +2395,17 @@ def contrastive_search(
             streamer.end()
 
         if return_dict_in_generate:
+            # Contrastive search works by forward looking at the next token, so we need to exclude it from
+            # `past_key_values` to be consistent with the other decoding methods
+            if model_kwargs.get("past_key_values") is not None:
+                past_key_values = []
+                for layer in model_kwargs["past_key_values"]:
+                    layer_past_key_values = []
+                    for item in layer:
+                        layer_past_key_values.append(item[..., :-1, :])
```

Comment:
Guaranteed consistent item shape? No. Guaranteed that the second to last dimension is the sequence length? Yes :) 

Some models squash together or permutate the first two dimensions (by default, the cache is `(batch_size, num_heads, sequence_length, embed_size_per_head)`), Bloom and GPTBigCode aka StarCoder being the biggest offenders.

One of the tests I added is in the mixin and touches contrastive search, so the fact that this output can be correctly used for continuations is tested :D In fact, I only noticed that this was a problem precisely because of the test!